SilKit/source/util/ExecutionEnvironment.cpp

```diff
@@ -0,0 +1,281 @@
+// SPDX-FileCopyrightText: 2024 Vector Informatik GmbH
+//
+// SPDX-License-Identifier: MIT
+
+#include "ExecutionEnvironment.hpp"
+
+#include "Assert.hpp"
+#include "FileHelpers.hpp"
+
+#include <array>
+#include <string>
+#include <vector>
+
+#include "asio.hpp"
+
+#include "fmt/format.h"
+
+#ifdef __unix__
+#include <unistd.h>
+#include <sys/utsname.h>
+#endif
+
+#ifdef __linux__
+#include <sys/sysinfo.h>
+#include <sys/types.h>
+#include <pwd.h>
+#endif
+
+#ifdef _WIN32
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+#include <lmcons.h>
+#include <psapi.h>
+#endif
+
+namespace VSilKit {
+
+namespace {
+
+constexpr const char* UNKNOWN_VALUE = "<unknown>";
+
+#ifdef __unix__
+
+auto GetUsername() -> std::string
+{
+    static const auto result = []() -> std::string {
+        const auto uid = ::getuid();
+        const auto pwd = ::getpwuid(uid);
+        if (pwd == nullptr)
+        {
+            return UNKNOWN_VALUE;
+        }
+
+        return pwd->pw_name;
+    }();
+
+    return result;
+}
+
+#ifdef __linux__
+
+auto GetProcessExecutable() -> std::string
+{
+    static const auto result = []() -> std::string {
+        std::vector<char> buffer;
+        buffer.resize(4096, '\0');
+
+        auto pathLength = ::readlink("/proc/self/exe", buffer.data(), buffer.size() - 1);
+        if (pathLength <= 0)
+        {
+            return UNKNOWN_VALUE;
+        }
+
+        buffer.resize(static_cast<std::size_t>(pathLength));
+        return std::string{buffer.begin(), buffer.end()};
+    }();
+
+    return result;
+}
+
+#else
+
+auto GetProcessExecutable() -> std::string
+{
+    return UNKNOWN_VALUE;
+}
+
+#endif
+
+auto GetProcessorCount() -> std::string
+{
+    static const auto result = std::to_string(::get_nprocs());
+    return result;
+}
+
+auto GetPageSize() -> std::string
+{
+    static const auto result = std::to_string(::sysconf(_SC_PAGESIZE));
+    return result;
+}
+
+#ifdef __linux__
+
+auto GetPhysicalMemoryMB() -> std::string
+{
+    static const auto result = [] {
+        const auto pageSize = static_cast<std::uint64_t>(::sysconf(_SC_PAGESIZE));
```

Comment:
sysconf is as posix as it gets. this could be inside a `#if __unix__`, I think

Reply:
It is inside of the `#ifdef __unix__`, just hard to see. Maybe I should place them around each function individually ...

Reply:
I restructured the entire file to make it (hopefully) much more readable!